### PR TITLE
feat: replace gltf sector parser with custom made parser

### PR DESCRIPTION
**Pull request by christjt**
This was done in order to improve performance. Chrome performance profiling shows  about a 5x performance increase in parsing sectors. (this will of course vary depending on the sector and its contents).

~~NB: instanced triangle meshes are not included (work in progress, and will be a separate PR).~~

### CODE REVIEW

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 1]
```diff
@@ -2,224 +2,289 @@
...
   }
 
-  public parseSector(data: ArrayBuffer): TypedGeometry[] {
-    const document: Document = this._gltfReader.readBinary(data);
+  public parseSector(data: ArrayBuffer) {
+    const headers = this._glbMetadataParser.parseGlbMetadata(data);
 
-    const defaultScene = document.getRoot().getDefaultScene();
-    const typedGeometryBuffers: TypedGeometry[] = [];
+    const json = headers.json;
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:52]:
```suggestion
```

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 2]
```diff
@@ -2,224 +2,289 @@
...
   constructor() {
-    this._gltfReader = new NodeIO();
-    this._gltfReader.registerExtensions([MeshGPUInstancing]);
+    this._glbMetadataParser = new GlbMetadataParser();
   }
 
-  public parseSector(data: ArrayBuffer): TypedGeometry[] {
-    const document: Document = this._gltfReader.readBinary(data);
+  public parseSector(data: ArrayBuffer) {
+    const headers = this._glbMetadataParser.parseGlbMetadata(data);
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:50]:
```suggestion
```

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 3]
```diff
@@ -2,224 +2,289 @@
...
+    const defaultSceneNodeIds = json.scenes[json.scene].nodes;
+
+    defaultSceneNodeIds
+      .map(nodeId => json.nodes[nodeId])
+      .forEach(node => {
+        const processedNode = this.processNode(node, headers, data)!;
+
+        if (processedNode === undefined) {
+          return;
+        }
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:68]:
```suggestion
        const processedNode = this.processNode(node, headers, data)!;
        if (processedNode === undefined) {
          return;
        }
```

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 4]
```diff
@@ -2,224 +2,289 @@
...
+      glbHeaderData,
+      instancingExtension: instancingExtension,
+      meshId,
+      data
+    };
 
     switch (geometryType) {
       case RevealGeometryCollectionType.InstanceMesh:
-        instanceId = this.processInstancedMesh(mesh, instanced, bufferGeometry);
+        //   instanceId = this.processInstancedMesh(mesh, instanced, bufferGeometry);
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:97]: I don't think it's a good idea to merge this PR before we have instance meshes and feature parity with as is (?)

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 5]
```diff
@@ -2,224 +2,289 @@
...
-        accessor.getElementSize()
-      );
+    const json = glbHeaderData.json;
 
-      bufferGeometry.setAttribute(`a${attributeName}`, instancedBufferAttribute);
-    });
-  }
+    const uniqueBufferViews = [
+      ...new Set(Object.values(attributes).map(accessorId => json.accessors[accessorId].bufferView))
+    ];
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:209]: Not 100% sure on whats going on here, but isn't it a bit excessive to create a Set if theres only one element?

File: viewer/packages/sector-parser/src/GltfSectorParser.ts [hunk 6]
```diff
@@ -2,224 +2,289 @@
...
+
+      assert(elementSize !== undefined);
+
+      const interleavedBufferAttribute = new THREE.InterleavedBufferAttribute(
+        interleavedBuffer,
+        size,
+        (accessor.byteOffset ?? 0) / elementSize
+      );
+      const transformedAttributeName = transformAttributeName(attributeName);
+      bufferGeometry.setAttribute(transformedAttributeName, interleavedBufferAttribute);
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/GltfSectorParser.ts:260]:
```suggestion
      const interleavedBuffer = typedArrayMap[accessor.componentType];
      const size = this.COLLECTION_TYPE_SIZES.get(accessor.type);
      assert(size !== undefined);
      const elementType = this.DATA_TYPE_BYTE_SIZES.get(accessor.componentType);
      assert(elementType !== undefined);
      const elementSize = elementType.BYTES_PER_ELEMENT;
      assert(elementSize !== undefined);

      const interleavedBufferAttribute = new THREE.InterleavedBufferAttribute(
        interleavedBuffer,
        size,
        (accessor.byteOffset ?? 0) / elementSize
      );
      const transformedAttributeName = transformAttributeName(attributeName);
      bufferGeometry.setAttribute(transformedAttributeName, interleavedBufferAttribute);
```
To me, it doesn't make the code more readable with all of these blank lines, but just a suggestion

File: viewer/packages/sector-parser/src/reveal-glb-parser/GlbMetadataParser.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+import assert from 'assert';
+import { GlbHeaderData, GltfJson } from '../types';
+
+export class GlbMetadataParser {
+  private _textDecoder: TextDecoder;
+
+  private readonly GLB_HEADER_BYTE_SIZE = 12;
+  private readonly CHUNK_HEADER_BYTE_SIZE = 8;
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/reveal-glb-parser/GlbMetadataParser.ts:12]: Make static

File: viewer/packages/sector-parser/src/reveal-glb-parser/GlbMetadataParser.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
+/*!
+ * Copyright 2021 Cognite AS
+ */
+
+import assert from 'assert';
+import { GlbHeaderData, GltfJson } from '../types';
+
+export class GlbMetadataParser {
+  private _textDecoder: TextDecoder;
```
Comment (larsmoa) [inline viewer/packages/sector-parser/src/reveal-glb-parser/GlbMetadataParser.ts:9]:
```suggestion
  private readonly _textDecoder: TextDecoder;
```

File: viewer/packages/sector-parser/tests/GltfSectorParser.test.ts [hunk 9]
```diff
@@ -5,7 +5,8 @@
 import fs from 'fs';
 
 import 'jest-extended';
-import GltfSectorParser, { RevealGeometryCollectionType } from '../src/GltfSectorParser';
+import { GltfSectorParser } from '../src/GltfSectorParser';
+import { RevealGeometryCollectionType } from '../src/types';
 
 describe('GltfSectorParser Tests', () => {
```
Comment (larsmoa) [inline viewer/packages/sector-parser/tests/GltfSectorParser.test.ts:11]:
```suggestion
describe('GltfSectorParser', () => {
```
I think we should keep the naming scheme we have or switch to a new for all tests and not use several. Also, I've been thinking of if we should do this instead:
```
describe(GltfSectorParser.name, () => {
```
Which would avoid names getting out of sync.

**Review by larsmoa — CHANGES_REQUESTED** [review]
Looks good, see some comments. 

However, I'm skeptical to merging before we have instance meshes - why not wait? Essentially we'd break existing functionality if we brought this in (even if its not master)

File: viewer/packages/sector-parser/src/GltfSectorParser.ts (hunk 5 above)
Comment (christjt) [reply viewer/packages/sector-parser/src/GltfSectorParser.ts:209]: Yeah, wasn't 100% satisfied with this. What I am trying to do is get all unique values of a given list of numbers. Basically want Linq `Distinct` method here. Any suggestions?

Comment (haakonflatval-cognite) [reply viewer/packages/sector-parser/src/GltfSectorParser.ts:209]: From what I can understand, `Object.values()` returns a list. It is funneled into a `Set` for deduplication

File: viewer/packages/sector-parser/src/GltfSectorParser.ts (hunk 6 above)
Comment (christjt) [reply viewer/packages/sector-parser/src/GltfSectorParser.ts:260]: I don't agree with this, and prefer the space when reading. Not obsessed with this though so can change if you are very opinionated about this.

File: viewer/packages/sector-parser/src/GltfSectorParser.ts (hunk 5 above)
Comment (larsmoa) [reply viewer/packages/sector-parser/src/GltfSectorParser.ts:205]: All you are trying to do is to ensure that all values are equal right? In this case you'd only need to iterate through them and ensure they are all equal to the first element.

**Review by larsmoa — APPROVED** [review]

**Pull request merged**